Re-implement handling of `RolloverCompleted` for taker

thomaseizinger opened the pull request.

thomaseizinger committed:
Introduce type alias for Completed<CollaborativeSettlement>

This makes it easier to introduce additional type parameters to
`Completed`.

thomaseizinger committed:
Introduce configurable error type for `Completed`

thomaseizinger committed:
Be more specific about handling the maker response timeout

Instead of failing the entire protocol after X amount of time, only
check whether our proposal got accepted by the maker within our timeout.
If it did, continue on. If it didn't fail the protocol. Not only should
this give us a much quicker response time in case the maker does not
response, it also more resilient in case rollover takes longer for
whatever reason.

Note that we still have timeout on the individual messages within
rollover anyway.

thomaseizinger committed:
Improve error handling when forwarding messages

thomaseizinger committed:
Make `update_proposal` non-fallible

thomaseizinger committed:
Make private what can be private

thomaseizinger committed:
Don't use spawn_fallible for the rollover task

The only error that can occur here is "Actor is disconnected". That can
only happen if we discard the rollover actor before the protocol
finishes. In that case, we are simply no longer interested in the result
and we can thus safely ignore this error.

To make sure that this thinking is not invalidated, we add an explicit
type annotation that enforces that we are not dropping an error here in
case someone ever changes the return value from the handler to NOT be a
`()` Ok value.

thomaseizinger committed:
Expose strongly typed rollover errors

thomaseizinger committed:
Rename function as per naming convention of `rollover`

thomaseizinger committed:
Inline pointless function

thomaseizinger committed:
Handle RolloverCompleted for taker in auto_rollover

da-kami commented on daemon/src/collab_settlement_maker.rs:
```diff
@@ -18,7 +17,7 @@ use xtra_productivity::xtra_productivity;
     cfd: Cfd,
     projection: xtra::Address<projection::Actor>,
-    on_completed: Box<dyn MessageChannel<Completed<CollaborativeSettlement>>>,
+    on_completed: Box<dyn MessageChannel<CollaborativeSettlementCompleted>>,
```
Note: I think the creation of the alias is somehow not part of this commit although it should be. This just shows the usage of the alias but not the alias itself.

Sidenote: 
We had this alias before, but I thought we agreed that we don't need to introduce one for collaborative settlement 😅

thomaseizinger replied:
I reintroduced it because adding the error type everywhere was annoying :D

thomaseizinger replied:
Thanks for pointing out the messy commit, will fix!

da-kami commented on daemon/src/model/cfd.rs:
```diff
@@ -1414,16 +1415,16 @@ pub mod marker {
+pub type RolloverCompleted = Completed<(Dlc, marker::Rollover), anyhow::Error>;
 
-pub type CollaborativeSettlementCompleted = Completed<CollaborativeSettlement>;
+pub type CollaborativeSettlementCompleted = Completed<CollaborativeSettlement, anyhow::Error>;
```
ah, here it is :D

da-kami commented on daemon/src/rollover_taker.rs:
```diff
@@ -305,6 +308,30 @@ impl Actor {
         .await;
     }
 
+    pub async fn handle_rollover_timeout_reached(
```
👍

da-kami commented on daemon/src/auto_rollover.rs:
```diff
@@ -112,8 +115,43 @@ where
+            _ => bail!("Unexpected event {:?}", event.event),
+        };
+
+        tracing::info!("Setup complete, publishing on chain now");
```
```suggestion
        tracing::info!("Rollover completed");
```

This message is a copy-paste from contract-setup and incorrect in this context.

thomaseizinger replied:
Thanks for pointing it out!

luckysori commented on daemon/src/rollover_taker.rs:
```diff
@@ -176,6 +177,11 @@ impl Actor {
+
+    /// Returns whether the maker has accepted our rollover proposal.
+    fn is_accepted(&self) -> bool {
+        self.rollover_msg_sender.is_some()
```
Hehe, nice heuristic.

luckysori commented on daemon/src/auto_rollover.rs:
```diff
@@ -139,10 +139,7 @@ where
-                rollover_taker::MAX_ROLLOVER_DURATION + Duration::from_secs(60),
-                || AutoRollover,
-            )
+            .notify_interval(Duration::from_secs(5 * 60), || AutoRollover)
```
❓ Why the magic number `5`?

thomaseizinger replied:
This just restores the previous functionality. `MAX_ROLLOVER_DURATION` was 4 minutes, I just made this not depend on this constant anymore because the internals now work differently.

luckysori commented on daemon/src/rollover_taker.rs:
```diff
@@ -160,18 +159,18 @@ impl Actor {
-        proposal: Option<(RolloverProposal, SettlementKind)>,
-    ) -> Result<()> {
-        self.projection
+    /// Update the proposal state in the projection actor.
+    ///
+    /// This function is non-fallible because we don't care if the actor goes away, it is not
+    /// critical to continue with the protocol.
```
👍 Good learning for the rest of the ephemeral actors.

thomaseizinger replied:
Well, the projection actor should likely be gone from these actors completely!

luckysori replied:
That is the dream!

luckysori commented on daemon/src/rollover_taker.rs:
```diff
@@ -137,7 +137,8 @@ impl Actor {
         let this = ctx.address().expect("self to be alive");
         self.tasks.add(async move {
-            // Use an explicit type annotation to cause a compile error if someone changes the handler.
+            // Use an explicit type annotation to cause a compile error if someone changes the
+            // handler.
```
🙃 Mega-nit, but this belongs in the previous commit.

luckysori reviewed (approved):
Very easy to review with all the small commits. LGTM!

thomaseizinger committed:
Rename field to match type

The type only knows about the `Completed` message, not which actor will
receive it.

thomaseizinger committed:
Make internal messages private

thomaseizinger committed:
Inline `complete` function

`on_completed` being disconnected just means nobody is interested in the
result of the protocol, that is an error we can safely ignore.

thomaseizinger committed:
Remove duplicate error message

The error returned from `oracle::Actor` says exactly the same thing.

thomaseizinger committed:
Align error handling of sender with `rollover_taker`

thomaseizinger committed:
Align `rollover_maker` with `rollover_taker`

thomaseizinger committed:
Shorten field names

thomaseizinger committed:
Handle `RolloverCompleted` in `maker_cfd::Actor`
